### PR TITLE
fix issue 54153 by not testing issue-18804 on Windows nor OS X.

### LINKED ISSUE #54153: Issue 18804 (linkage weak_extern through generics) regressing on OS X?
I am locally seeing test failures on OS X 

```
---- [ui] ui/run-pass/issues/issue-18804/main.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit code: 1
command: "/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/stage1/bin/rustc" "/Users/fklock/Dev/Mozilla/issue53528/rust-53528/src/test/ui/run-pass/issues/issue-18804/main.rs" "--target=x86_64-apple-darwin" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "-o" "/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/a" "-Crpath" "-Zunstable-options" "-Lnative=/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/native/rust-test-helpers" "-L" "/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/auxiliary" "-A" "unused"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
{"message":"linking with `cc` failed: exit code: 1","code":null,"level":"error","spans":[],"children":[{"message":"\"cc\" \"-m64\" \"-L\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/stage1/lib/rustlib/x86_64-apple-darwin/lib\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/a.main.7rcbfp3g-cgu.0.rcgu.o\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/a.main.7rcbfp3g-cgu.1.rcgu.o\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/a.main.7rcbfp3g-cgu.2.rcgu.o\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/a.main.7rcbfp3g-cgu.3.rcgu.o\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/a.main.7rcbfp3g-cgu.4.rcgu.o\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/a.main.7rcbfp3g-cgu.5.rcgu.o\" \"-o\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/a\" \"-Wl,-dead_strip\" \"-nodefaultlibs\" \"-L\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/native/rust-test-helpers\" \"-L\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/auxiliary\" \"-L\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/stage1/lib/rustlib/x86_64-apple-darwin/lib\" \"-L\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/auxiliary\" \"-llib\" \"-L\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/stage1/lib/rustlib/x86_64-apple-darwin/lib\" \"-lstd-bc4d93c7cd68669e\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/stage1/lib/rustlib/x86_64-apple-darwin/lib/libcompiler_builtins-4219949d54044b85.rlib\" \"-lSystem\" \"-lresolv\" \"-lpthread\" \"-lc\" \"-lm\" \"-Wl,-rpath,@loader_path/auxiliary\" \"-Wl,-rpath,@loader_path/../../../../../../stage1/lib/rustlib/x86_64-apple-darwin/lib\" \"-Wl,-rpath,/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/~/opt/rust-dbg/lib/rustlib/x86_64-apple-darwin/lib\"","code":null,"level":"note","spans":[],"children":[],"rendered":null},{"message":"Undefined symbols for architecture x86_64:\n  \"_FOO\", referenced from:\n      __rust_extern_with_linkage_FOO in a.main.7rcbfp3g-cgu.4.rcgu.o\nld: symbol(s) not found for architecture x86_64\nclang: error: linker command failed with exit code 1 (use -v to see invocation)\n","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error: linking with `cc` failed: exit code: 1\n   |\n   = note: \"cc\" \"-m64\" \"-L\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/stage1/lib/rustlib/x86_64-apple-darwin/lib\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/a.main.7rcbfp3g-cgu.0.rcgu.o\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/a.main.7rcbfp3g-cgu.1.rcgu.o\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/a.main.7rcbfp3g-cgu.2.rcgu.o\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/a.main.7rcbfp3g-cgu.3.rcgu.o\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/a.main.7rcbfp3g-cgu.4.rcgu.o\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/a.main.7rcbfp3g-cgu.5.rcgu.o\" \"-o\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/a\" \"-Wl,-dead_strip\" \"-nodefaultlibs\" \"-L\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/native/rust-test-helpers\" \"-L\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/auxiliary\" \"-L\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/stage1/lib/rustlib/x86_64-apple-darwin/lib\" \"-L\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/test/ui/run-pass/issues/issue-18804/main/auxiliary\" \"-llib\" \"-L\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/stage1/lib/rustlib/x86_64-apple-darwin/lib\" \"-lstd-bc4d93c7cd68669e\" \"/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/build/x86_64-apple-darwin/stage1/lib/rustlib/x86_64-apple-darwin/lib/libcompiler_builtins-4219949d54044b85.rlib\" \"-lSystem\" \"-lresolv\" \"-lpthread\" \"-lc\" \"-lm\" \"-Wl,-rpath,@loader_path/auxiliary\" \"-Wl,-rpath,@loader_path/../../../../../../stage1/lib/rustlib/x86_64-apple-darwin/lib\" \"-Wl,-rpath,/Users/fklock/Dev/Mozilla/issue53528/rust-53528/objdir-dbgopt/~/opt/rust-dbg/lib/rustlib/x86_64-apple-darwin/lib\"\n   = note: Undefined symbols for architecture x86_64:\n             \"_FOO\", referenced from:\n                 __rust_extern_with_linkage_FOO in a.main.7rcbfp3g-cgu.4.rcgu.o\n           ld: symbol(s) not found for architecture x86_64\n           clang: error: linker command failed with exit code 1 (use -v to see invocation)\n           \n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}

------------------------------------------

thread '[ui] ui/run-pass/issues/issue-18804/main.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3196:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.


failures:
    [ui] ui/run-pass/issues/issue-18804/main.rs
```

I am not certain what is happening yet. After reviewing #18804, I have looked briefly at the generated LLVM IR output for main.rs, and it seems like `FOO` is being properly tagged:

```
%"unwind::libunwind::_Unwind_Exception" = type { [0 x i64], i64, [0 x i64], void (i32, %"unwind::libunwind::_Unwind_Exception"*)*, [0 x i64], [6 x i64], [0 x i64] }
%"unwind::libunwind::_Unwind_Context" = type { [0 x i8] }

@FOO = extern_weak global {}
@_rust_extern_with_linkage_FOO = internal global {}* @FOO
```

### PR DESCRIPTION
**Pull request by pnkfelix**
Fix #54153

### CODE REVIEW

File: src/test/run-pass/issues/issue-18804/main.rs [hunk 1]
```diff
@@ -17,6 +17,9 @@
 
 // aux-build:lib.rs
 
+// rust-lang/rust#54143: This test *requires* optimizations to be turned on.
```
Comment (nikic) [inline src/test/run-pass/issues/issue-18804/main.rs]:
```suggestion
// rust-lang/rust#54153: This test *requires* optimizations to be turned on.
```

**Comment by nikic** [comment]
It's not entirely obvious to me why this test requires optimization.

File: src/test/run-pass/issues/issue-18804/main.rs (hunk 1 above)
Comment (pnkfelix) [reply src/test/run-pass/issues/issue-18804/main.rs]: ugh missed that one, thanks

**Comment by pnkfelix** [comment]
its not obvious to me either; the requirement may only manifest itself on "OS X" ...

which would lead one to wonder if instead I should tag this test with an `// ignore-macos`, and then file an additional test just of "OS X" that has the `-O`  flag required?

In any case I am non-plussed that I cannot do `optimize-tests = false` in my `config.toml` without running into this.

**Comment by pnkfelix** [comment]
cc #18804

**Comment by nikic** [comment]
On Ubuntu, at least a manual run seems to work fine. I did:

```
rustc +nightly -C opt-level=0 lib.rs
rustc +nightly --extern lib=liblib.rlib -C opt-level=0 test.rs
./test
```

Not sure how well that corresponds to whatever run-pass does.

**Comment by alexcrichton** [comment]
Ah yeah I agree with @nikic, this test should be ignored on macos/windows because the `extern_weak` linkage doesn't really exist there

**Comment by pnkfelix** [comment]
Okay I'll revise the PR.

**Comment by nikic** [comment]
@bors r-

Failed in the rollup with ... an LLVM assertion failure :(

```
[01:00:08] ---- [run-pass] run-pass/issues/issue-18804/main.rs stdout ----
[01:00:08] 
[01:00:08] error: test compilation failed although it shouldn't!
[01:00:08] status: signal: 6
[01:00:08] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-pass/issues/issue-18804/main.rs" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/issues/issue-18804/main/a" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-C" "no-prepopulate-passes" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/issues/issue-18804/main/auxiliary"
[01:00:08] ------------------------------------------
[01:00:08] 
[01:00:08] ------------------------------------------
[01:00:08] stderr:
[01:00:08] stderr:
[01:00:08] ------------------------------------------
[01:00:08] rustc: /checkout/src/llvm/include/llvm/ADT/StringRef.h:239: char llvm::StringRef::operator[](size_t) const: Assertion `Index < Length && "Invalid index!"' failed.
[01:00:08] ------------------------------------------
[01:00:08] 
[01:00:08] thread '[run-pass] run-pass/issues/issue-18804/main.rs' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:3252:9
[01:00:08] note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by nikic** [comment]
Just checked this locally with an assertion-enabled build ... the issue is that due to `-C no-prepopulate-passes` the name-anon-globals pass is not run, but we have anon globals and are using ThinLTO buffers. We have checks in place to emit a compile error in this case (rather than this wonderful LLVM assertion), but they happen to miss this particular case.

To fix this you can either drop the `-C no-prepopulate-passes` or add an additional `-C passes=name-anon-globals`.

We should probably start always running `name-anon-globals` (even under `no-prepopulate-passes`), as this issue has come up a few times now and we never quite seem to catch all the cases where it's necessary to run it.

**Pull request merged**